File: sources/CSharp/src/Ers/Visualization/RenderFramework.cs
using Ers.Engine;

namespace Ers
{
    public class RenderFramework
    {
        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public IntPtr CorePtr;

        public RenderFramework(IntPtr windowHandle, IntPtr displayHandle, int width, int height)
        {
            this.CorePtr = ErsEngine.ERS_RenderFramework_Create(windowHandle, displayHandle, width, height);
        }

        public void DrawRenderContext(RenderContext renderContext)
        {
            ErsEngine.ERS_RenderFramework_DrawRenderContext(CorePtr, renderContext.CorePtr);
        }

        public void Present() => ErsEngine.ERS_RenderFramework_Present(CorePtr);

        public void DisposeInner()
        {
            if (CorePtr != IntPtr.Zero)
            {
                ErsEngine.ERS_RenderFramework_Destroy(CorePtr);
                CorePtr = IntPtr.Zero;
            }
        }

        ~RenderFramework() => DisposeInner();
    }
}

File: sources/CPP/include/Ers/Visualization/RenderFramework.cpp
#include "RenderFramework.h"
#include "Ers/Api.h"

namespace Ers
{
    RenderFramework::RenderFramework(void* windowHandle, void* displayHandle, int screenWidth, int screenHeight)
    {
        corePtr = Ers::Engine::ERS_RenderFramework_Create(windowHandle, displayHandle, screenWidth, screenHeight);
    }

    void RenderFramework::Present()
    {
        Ers::Engine::ERS_RenderFramework_Present(corePtr);
    }

    void RenderFramework::DrawRenderContext(RenderContext& renderContext)
    {
        Ers::Engine::ERS_RenderFramework_DrawRenderContext(corePtr, renderContext.CorePtr());
    }

    RenderFramework::~RenderFramework()
    {
        Ers::Engine::ERS_RenderFramework_Destroy(corePtr);
    }

} // namespace Ers
